pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,530 @@
+            file_util::fput_multiple<T>(src, n_elements, fp_);
+        }
+
+        /* Save one element */
+        template<class T, class TT=T, if_simple_serializable<TT> = true>
+        void fput_one(const T& src) { // Call fput_multiple
+            fput_multiple<T>(&src, 1u);
+        }
+
+    private:
+        MmapMetadata_ metadata_; // Metadata
+        FILE* fp_; // File pointer
+
+        /* Save metadata to end of file after all data blocks have been saveed.
+         * Also, file pointer is closed at the end.
+         */
+        void finalize_() {
+            metadata_.save(fp_);
+            if (fclose(fp_) != 0) {
+                throw std::runtime_error("MmapStoreSave: Close file failed.");
+            }
+        }
+}; // end class MmapStoreSave_
+
+} // end namespace details_
+
+
+/*
+ * Wrapper of details_::MmapStoreSave_ and details_::MmapStoreLoad_ to facilitate usage
+ */
+class MmapStore {
+    public:
+        // Disable copy/move for single ownership semantics
+        MmapStore(const MmapStore&) = delete;
+        MmapStore(MmapStore&&);
+        MmapStore& operator=(const MmapStore&) = delete;
+        MmapStore& operator=(MmapStore&&);
+
+        /* Dummy Constructor */
+        MmapStore() { }
+        /* Destructor to delete any existing instance */
+        ~MmapStore() { close(); }
+
+        /* Open a filename with given mode read/write and options
+         * Parameters:
+         *      file_name (string): Path of the file
+         *      mode_str (string): Open for following options:
+         *          * read ("r") : pre-fault all pages
+         *          * read with lazy load ("r_lazy") : only create mmap, loading when accessing
+         *          * write ("w")
+         */
+        void open(const std::string& file_name, const std::string& mode_str) {
+            if (mode_ != Mode_::UNINIT) {
+                throw std::runtime_error("Should close existing file before open new one.");
+            }
+            if (mode_str == "r") { // pre-load all pages
+                mmap_r_ = new details_::MmapStoreLoad_(file_name, true);
+                mode_ = Mode_::READONLY;
+            }
+            else if (mode_str == "r_lazy") {
+                mmap_r_ = new details_::MmapStoreLoad_(file_name, false);
+                mode_ = Mode_::READONLY;
+            }
+            else if (mode_str == "w") {
+                mmap_w_ = new details_::MmapStoreSave_(file_name);
+                mode_ = Mode_::WRITEONLY;
+            }
+            else {
+                throw std::runtime_error("Unrecogonized mode. Should be either 'r', 'r_lazy' or 'w'.");
+            }
+        }
+
+        /* Close any opened files
+         * If open for read, should NOT close until the data of the file is not visited anymore.
+         */
+        void close() {
+            if (mode_ == Mode_::UNINIT) {
+                return;
+            }
+            else if (mode_ == Mode_::READONLY) {
```

Comment:
let's merge these two lines into one `} else if {`, which is used as a convection in other C++ codes in the library.